### ISSUE TITLE
Evaluation of the model is stochastic ?

### ISSUE THREAD

**Issue by MmelodYy** [UCSB-AI/Aerial-Vision-and-Dialog-Navigation#19]
Dear Author:

Hello!

Thank you very much for your wonderful work. I have a question to confirm with you. I found that the original code uses 'random.shuffle(self.data)' in env.py. This causes the results to be different every time when testing.

![bug](https://github.com/user-attachments/assets/7269affd-b0b9-47e8-adaf-95c6bb290dc5)

**Comment by MmelodYy** [comment]
After several checks, **I confirmed that this was indeed an error** and that it would affect the results of the model evaluation. Unless the **batch size** is set to **1** when testing the **val_seen,val_unseen,test_unseen** dataset.

Specifically, this is mainly due to the phrase **random.shuffle(self.data)** in the next_batch function. When the length of the data is not divisible by the **batch size**, the header data is used to fill the gap. But calling r**andom.shuffle(self.data)** before filling the gap will result in different data being used to fill the gap each time.

**As a result, the test results obtained by loop iteration during training are different. Therefore, the test results (on val_seen and val_unseen datasets) during the training process cannot be used as the final model results.** (I found that someone has already done this.) I have tried to use the final model weights and loop the test 10 times, and found that the results are different each time, and some of the metrics are not small differences. But when I set the batch size to 1 while testing, the data doesn't perform an end-fill operation, and then the results are stable each time.

To summarize, this bug causes the numerical results of the test to be incorrect. However, it should be noted that even though the program has a bug that causes a patch operation at the end of the test_unseen dataset, t**he relative size of the values on test_unseen is still informative because the random seed is fixed.**

**Comment by MmelodYy** [comment]
Here are some experimental evidences: (**action** represents each action step [calling the ' self.env._get_obs' function]; **evl_item** represents Simulated values of the ‘self.data’ index.)

exp1: batch size=1, use random.shuffle(self.data)
![1](https://github.com/user-attachments/assets/55f930c2-063e-43cf-9bd1-4f2676f7e91e)

exp2: batch size=4, without random.shuffle(self.data) == **for loop testing** == use different random seed
![4](https://github.com/user-attachments/assets/1b496d17-e3f7-4617-9b04-ec23c2d88db4)

exp3: batch size=4, use random.shuffle(self.data)
![30](https://github.com/user-attachments/assets/40212479-1c54-41a8-b347-005fee5e350f)

As we can see, **this will lead to the program using different data for model testing, resulting in different experimental results.**

**Comment by UeFan** [comment]
Hi, 

I confirm this is a bug. Thank you for bringing up this. I have just fixed it (https://github.com/eric-ai-lab/Aerial-Vision-and-Dialog-Navigation/commit/fe121351368949406365615b675c6e55f17be9b0). 

I just want to point out that I checked my previous experiment logs, and I found that I was always using batch size = 1 during the evaluation, so thankfully, this line of random shuffle was never used in my past evaluation, and the results reported in the paper should be reliable. 

Sorry for the inconvenience and don't hesitate to ask if you have more question.
